storage/internal/redis6/redis.go

```diff
@@ -0,0 +1,97 @@
+package redis
+
+import (
+	"crypto/tls"
+	"strconv"
+	"time"
+
+	"github.com/TykTechnologies/tyk/config"
+	logger "github.com/TykTechnologies/tyk/log"
+)
+
+// New returns a go-redis:v8 universal client
+func New(isCache, isAnalytics bool, conf config.Config) *Driver {
+	// redisSingletonMu is locked and we know the singleton is nil
+	cfg := conf.Storage
+	if isCache && conf.EnableSeperateCacheStore {
+		cfg = conf.CacheStorage
+	} else if isAnalytics && conf.EnableAnalytics && conf.EnableSeperateAnalyticsStore {
+		cfg = conf.AnalyticsStorage
+	}
+
+	log := logger.Get()
+	log.Debug("Creating new Redis connection pool")
+
+	// poolSize applies per cluster node and not for the whole cluster.
+	poolSize := 500
+	if cfg.MaxActive > 0 {
+		poolSize = cfg.MaxActive
+	}
+
+	timeout := 5 * time.Second
+
+	if cfg.Timeout > 0 {
+		timeout = time.Duration(cfg.Timeout) * time.Second
+	}
+
+	var tlsConfig *tls.Config
+
+	if cfg.UseSSL {
+		tlsConfig = &tls.Config{
+			InsecureSkipVerify: cfg.SSLInsecureSkipVerify,
+		}
+	}
+
+	var client UniversalClient
+	opts := &UniversalOptions{
+		Addrs:            getRedisAddrs(cfg),
+		MasterName:       cfg.MasterName,
+		SentinelPassword: cfg.SentinelPassword,
+		Username:         cfg.Username,
+		Password:         cfg.Password,
+		DB:               cfg.Database,
+		DialTimeout:      timeout,
+		ReadTimeout:      timeout,
+		WriteTimeout:     timeout,
+		IdleTimeout:      240 * timeout,
+		PoolSize:         poolSize,
+		TLSConfig:        tlsConfig,
+	}
+
+	if opts.MasterName != "" {
+		log.Info("--> [REDIS] Creating sentinel-backed failover client")
+		client = NewFailoverClient(opts.Failover())
+	} else if cfg.EnableCluster {
+		log.Info("--> [REDIS] Creating cluster client")
+		client = NewClusterClient(opts.Cluster())
+	} else {
+		log.Info("--> [REDIS] Creating single-node client")
+		client = NewClient(opts.Simple())
+	}
+
+	return NewDriver(client)
+}
+
+func getRedisAddrs(config config.StorageOptionsConf) (addrs []string) {
+	if len(config.Addrs) != 0 {
+		addrs = config.Addrs
+	} else {
+		for h, p := range config.Hosts {
+			addr := h + ":" + p
+			addrs = append(addrs, addr)
+		}
+	}
+
+	if len(addrs) == 0 && config.Port != 0 {
+		addr := config.Host + ":" + strconv.Itoa(config.Port)
+		addrs = append(addrs, addr)
+	}
+
+	return addrs
+}
+
+// PubSub v8 and v9 :D
+type PubSub interface {
+	// I bet myself a burger this won't work
+
```

Comment:
Putting review comment placeholder to not forget about it

Reply:
This has been replaced by model.PubSub, did indeed work, and now I owe myself a burger despite all my pessimism.